Refactor GooglePlacesUtils.getAddressComponent

bondydaa opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

Cleaning up the `getAddressComponents` lib so that it returns an default empty string so that when calling it we do not need to "null coalesce" it like we are currently doing like so here: https://github.com/Expensify/App/blob/e3029e538a185431111220b82550ba21628a28fc/src/components/AddressSearch.js#L92-L93

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/14324   


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console
1. Go to Settings > Payments <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973083-9d12ab3b-f90f-44ba-b242-363c48c797fc.png">
2. Click on Add Payment Method <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973126-58d37c18-ac87-44cb-a776-d93fcd6beef6.png">
3. Click debit card <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973194-e7080ee7-31ce-4aac-bcc6-cfe8142d8a04.png">
4. Click into the billing address text input, enter `232 stonemill` and select the Dayton OH option from the suggestion, ensure it fills in the zip code and state like so <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973422-13f95ba4-4b84-480d-94ba-0e6f8748dff0.png">
5. Click into the billing address text input again, remove `232 stonemill` and enter `texas beltway` and select the `texas 8 beltway houston` option from the suggestion, ensure the zip code is cleared out but TX shows as the State like so <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973564-ee655294-4b6a-4ee5-8f1d-8fe51423acb7.png">
6. Click into the billing address text input again and enter `88 Kearny`, select the San Francisco option from the suggestions and ensure the rest of the data populates like so: <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973699-5bcd50c6-19b1-4320-9258-88210ecb061f.png">

**Test Personal Details with Country Selector**
1. Open Settings > Profile > Personal Details
2. Click Home Address
3. Search `88 Kearny St` and click the San Francisco option that shows up <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/218216041-87dcad9e-08d0-4e8c-95c8-0d5a76c16742.png">
4. Ensure it autofills all these details: <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/218216156-d5f79e04-7f21-4b9f-8fec-3b9334e35543.png">
5. Search again for `Kathmandustraat` and click the Netherlands suggestion and ensure the Country is updated and all details match this: <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/218216469-2308d006-fb4d-4b59-8d16-6f761d4ba8e2.png">


### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

I tried to find a way to prevent the autocomplete from making any requests to avoid these console errors but I don't think it's possible. The library doesn't offer a "prevent request" prop. I found this line https://github.com/FaridSafi/react-native-google-places-autocomplete/blob/fb9d70b04cb6403348bd030816bccac2f534315d/GooglePlacesAutocomplete.js#L484 which I'd hoped if we did something like this: 
```diff
diff --git a/src/components/AddressSearch.js b/src/components/AddressSearch.js
index 880fe79587..289365e9df 100644
--- a/src/components/AddressSearch.js
+++ b/src/components/AddressSearch.js
@@ -10,6 +10,9 @@ import styles from '../styles/styles';
 import TextInput from './TextInput';
 import Log from '../libs/Log';
 import * as GooglePlacesUtils from '../libs/GooglePlacesUtils';
+import {withNetwork} from '../components/OnyxProvider';
+import networkPropTypes from '../components/networkPropTypes';
+import compose from '../libs/compose';
 
 // The error that's being thrown below will be ignored until we fork the
 // react-native-google-places-autocomplete repo and replace the
@@ -56,6 +59,9 @@ const propTypes = {
         zipCode: PropTypes.string,
     }),
 
+    /** Information about the network from Onyx */
+    network: networkPropTypes.isRequired,
+
     ...withLocalizePropTypes,
 };
 
@@ -195,6 +201,11 @@ const AddressSearch = (props) => {
                         onBlur: props.onBlur,
                         autoComplete: 'off',
                         onInputChange: (text) => {
+                            console.log(text);
+                            if (props.network.isOffline) {
+                                text = undefined;
+                            }
+                            console.log(text);
                             if (props.inputID) {
                                 props.onInputChange(text);
                             } else {
@@ -242,7 +253,10 @@ AddressSearch.propTypes = propTypes;
 AddressSearch.defaultProps = defaultProps;
 AddressSearch.displayName = 'AddressSearch';
 
-export default withLocalize(React.forwardRef((props, ref) => (
+export default compose(
+    withLocalize,
+    withNetwork(),
+)(React.forwardRef((props, ref) => (
     // eslint-disable-next-line react/jsx-props-no-spreading
     <AddressSearch {...props} innerRef={ref} />
 )));
```

It would prevent the `_request` but that `onInputChange` method seems to be run after the request is made because `text` is still send to `_request` instead of `undefined`.


<img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973925-fb6cd718-a21e-4f67-9d58-8c1a75d08c3a.png">


### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
8. Upload an image via copy paste
9. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console
1. Go to Settings > Payments <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973083-9d12ab3b-f90f-44ba-b242-363c48c797fc.png">
4. Click on Add Payment Method <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973126-58d37c18-ac87-44cb-a776-d93fcd6beef6.png">
6. Click debit card <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973194-e7080ee7-31ce-4aac-bcc6-cfe8142d8a04.png">
7. Click into the billing address text input, enter `232 stonemill` and select the Dayton OH option from the suggestion, ensure it fills in the zip code and state like so <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973422-13f95ba4-4b84-480d-94ba-0e6f8748dff0.png">
8. Click into the billing address text input again, remove `232 stonemill` and enter `texas beltway` and select the `texas 8 beltway houston` option from the suggestion, ensure the zip code is cleared out but TX shows as the State like so <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973564-ee655294-4b6a-4ee5-8f1d-8fe51423acb7.png">
9. Click into the billing address text input again and enter `88 Kearny`, select the San Francisco option from the suggestions and ensure the rest of the data populates like so: <img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973699-5bcd50c6-19b1-4320-9258-88210ecb061f.png">

Or you can follow the steps on https://github.com/Expensify/App/issues/14324

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214987664-8a763f08-761f-4149-8f0c-28852e9ad051.png">
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/214973699-5bcd50c6-19b1-4320-9258-88210ecb061f.png">


https://user-images.githubusercontent.com/4073354/216727234-18ac0570-4844-4bbc-8f9f-f54a1cd1de9e.mp4



</details>

<details>
<summary>Mobile Web - Chrome</summary>


https://user-images.githubusercontent.com/4073354/216727260-ca8175d6-41bc-42ce-82cc-594c5f39b0e3.mp4



</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/4073354/216727285-1bf89223-dbee-4a43-96c3-92ac2c94d8a1.mp4


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/4073354/216727328-8840eab0-54d3-4f6a-9e61-7e27388fe66c.mp4



</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->


https://user-images.githubusercontent.com/4073354/216727346-017af4ac-26cd-4751-8295-7de5410c593b.mp4



</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->


https://user-images.githubusercontent.com/4073354/216727367-8dbef31a-ecf5-4492-b41a-997aa98065e7.mp4



</details>

bondydaa committed:
add in proposal code, implement other solution and test performance

bondydaa committed:
update tests to ensure when this is called multiple ways it always works

bondydaa committed:
trying to remove loops with nested array iterator built-ins

bondydaa committed:
splitting code into separate methods to more easily capture performance metrics and trying to stop loops with multiple interators

bondydaa committed:
adding big objects to test how performance is with large objects

bondydaa committed:
fix log line

bondydaa committed:
comment out timing in method calls for easier viewing results

bondydaa committed:
call methods 100 times in a row and see which is faster

bondydaa committed:
update to test for smaller object execution time as well

bondydaa committed:
remove testing code and leave only code that performed the quickest for big and small data sets

bondydaa committed:
revert performance tests and only leave ones for method that works best, add tests to ensure we dont have massive regressions for execution of this method

bondydaa committed:
update addressSearch to use the new interface of the GooglePlacesLib

bondydaa committed:
dont use lodashGet

tgolen reviewed (commented):
@bondydaa You definitely peaked my interest with this one. While I don't 100% get all the data transformation that is happening, I was able to write a version of `getAddressComponents()` which is able to run 1000 times in 0.3ms.

The current code can run 1000 times in 9ms. 

This is an increase in performance of 2900% 😁 

Now, I know my code isn't 100% correct, because the test for `should find address components by type` is failing one of the checks, but maybe you can spot the error and still use my approach.

```
    const result = {};
    for (let i = addressComponents.length - 1; i > 0; i--) {
        const addressComponent = addressComponents[i];
        for (let j = addressComponent.types.length - 1; j > 0; j--) {
            const addressType = addressComponent.types[j];
            if (addressComponent[fieldsToExtract[addressType]]) {
                result[addressType] = addressComponent[fieldsToExtract[addressType]];
            }
        }
    }
```

bondydaa committed:
Merge branch 'main' of github.com:Expensify/App into bondy-refactor-address-search

bondydaa commented:
ah you had an off-by-one because of the `i > 0` and `j > 0` so it wasn't looping the whole array, they need to be `i >= 0` `j >=0` (or `> -1`).

Only other thing then is that we're not adding the default empty string in yours either but noodling on that now.

tgolen commented:
OK, this is the part that I think it needs to stay away from and is adding a lot of unnecessary logic.

```
const result = Object.keys(fieldsToExtract).reduce((obj, field) => ({...obj, [field]:''}), {});
```

I think all you're trying to do is default the result value to an empty string, which you can do inside the loops like this?

```
    const result = {};
    for (let i = addressComponents.length - 1; i >= 0; i--) {
        const addressComponent = addressComponents[i];
        for (let j = addressComponent.types.length - 1; j >= 0; j--) {
            const addressType = addressComponent.types[j];
            result[addressType] = addressComponent[fieldsToExtract[addressType]] ? addressComponent[fieldsToExtract[addressType]] : '';
        }
    }
```

bondydaa commented:
That adds every key from `addressComponents` into the return list when we only want `fieldsToExtract`. 

Though maybe that's not the end of the world either 🤔

bondydaa committed:
remove unused imports

bondydaa committed:
simplify setting value

bondydaa committed:
fix styles

bondydaa commented:
okay publishing this now that I think we're all in agreement for using a little less performant but still fast code. working on finishing off checklist / screenshots.

bondydaa committed:
Merge branch 'main' of github.com:Expensify/App into bondy-refactor-address-search

bondydaa committed:
this seems to take long on GH actions so bump it up a little

bondydaa committed:
fix comment explaining arbitrary time picking

joelbettner commented on src/libs/GooglePlacesUtils.js:
```diff
@@ -10,18 +10,24 @@ import _ from 'underscore';
+            result[addressType] = value;
+        });
+    });
+    return result;
```
I'm not sure I understand what is happening here...
If we look at the example from the documentation above and the application in ApplicationSearch.js

```js
GooglePlacesUtils.getAddressComponents({long_name: "New York", short_name: "New York", types: [ "locality", "political" ]}, {
            street_number: 'long_name',
            route: 'long_name',
            locality: 'long_name',
            sublocality: 'long_name',
            postal_code: 'long_name',
            administrative_area_level_1: 'short_name',
        });
```

Then the `result` just becomes an array of empty strings that is the same size as whatever we put in as the second argument and we never update it to anything else...
![image](https://user-images.githubusercontent.com/13264774/217372710-cc057a94-5352-42ae-a04f-9224496eddb1.png)

joelbettner commented on src/libs/GooglePlacesUtils.js:
```diff
@@ -10,18 +10,24 @@ import _ from 'underscore';
- * @param {String} type
- * @param {String} key
- * @returns {String|undefined}
+ * @param {Object} fieldsToExtract – has shape: {addressType: 'keyToUse'}
```
This does not look like what we are passing in from AddressSearch.js...
```js
{
            street_number: 'long_name',
            route: 'long_name',
            locality: 'long_name',
            sublocality: 'long_name',
            postal_code: 'long_name',
            administrative_area_level_1: 'short_name',
}
```

joelbettner replied:
Spoke about this 1:1. It appears the documentation for the shape of `fieldsToExtract` isn't correct (the object should be wrapped in an array). Tested it after that and it worked.

bondydaa replied:
updated the comment above, `fieldToExtract` should be right, right?

joelbettner reviewed (changes requested):
Just update the doc block to show that the `addressComponents` argument is passed as an array of objects rather than just an object.

bondydaa committed:
update structure of example addressComponents

bondydaa committed:
fixing merge conflict

bondydaa committed:
add test for country google place look up

bondydaa commented:
fixed merged conflicts due to this PR https://github.com/Expensify/App/pull/14290/ which added the `country` field so I updated my test steps with screen shots how to test that and also updated the unit tests. 

Should be ready for another round of reviews 🙏

bondydaa committed:
fix destructuring style

mananjadhav commented:
I reviewed the PR and currently testing it. I tested on web platform and it worked fine. Will test out other platforms and complete the checklist.

mananjadhav commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/3069065/218535758-00040b63-8595-4f4f-939a-61d71532d3db.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/3069065/218536044-e3735c1d-3dcf-4e58-844f-3873388c0108.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/3069065/218536400-dcbbdfe7-38b9-4db6-9a15-c558e93cef70.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/3069065/218536436-4ebe8fe9-480b-4cf8-87fe-ba8031eec678.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/3069065/218536474-13e24886-2edf-4356-8cbc-e48b1a9ef324.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/3069065/218537945-f52c2792-2a73-4f63-a0f7-d940b43b5e26.mov

</details>


-----

Thanks for the patience here. I don't have a fake debit card info to fully submit the form. But I tested the Address and Google places part, it works fine. Thank you @bondydaa for the PR.

roryabraham reviewed (approved):
Haven't tested beyond the automated tests I wrote connected to the issue, but this code LGTM

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4186486659) to staging by https://github.com/roryabraham in version: 1.2.72-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4198237138) to production by https://github.com/francoisl in version: 1.2.72-1 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
